### PR TITLE
Linguist tags processor

### LINKED ISSUE #18258: 🚀 Feature: Catalog filter by language tag, powered by linguist plugin
### 🔖 Feature description

Allow for the [linguist plugin](https://github.com/backstage/backstage/tree/master/plugins/linguist) to emit entity tags, allowing for users to easily search for components by language in the catalog index page.

### 🎤 Context

Often times, people ask a question like "How many services do we have written in NodeJS vs Java". We could ask for teams to self-report by manually adding tags, but automation may provide more value more quickly.

### ✌️ Possible Implementation

Linguist plugin provides a processor you can hook up to the catalog which will fetch the language breakdown of a component and then modify the entity to include the language tag if it is not already there.

Likely we would also want this configurable, for example:

- List of allowed languages to add tags for
- List of disallowed languages to not add tags for
- Number of bytes threshold for a language to exceed before the tag is added for that language

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

### PR DESCRIPTION
**Pull request by bforbis**
## Hey, I just made a Pull Request!

This adds a processor to the linguist backend which provides the ability to use the languages breakdown from linguist and automatically add them as entity tags.

Closes #18258

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by pjungermann** [comment]
@awanlin FYI

File: plugins/linguist-backend/README.md [hunk 1]
```diff
@@ -53,10 +53,25 @@ Here's how to get the backend up and running:
      apiRouter.use('/linguist', await linguist(linguistEnv));
    ```
 
-4. Now run `yarn start-backend` from the repo root
-5. Finally open `http://localhost:7007/api/linguist/health` in a browser and it should return `{"status":"ok"}`
+4. Optionally, if you would like to have the linguist backend supply language tags to your entities so they can be filtered on in the catalog index, setup the linguist tag processor in `packages/backend/src/plugins/catalog.ts`.
```
Comment (awanlin) [inline plugins/linguist-backend/README.md:56]: I'd prefer having this as it's own section afterwards seeing as it's optional. This makes it easier for those already using Linguist to just jump to the new section and for those setting it up they can get that done and then add on this is they want it.

File: plugins/linguist-backend/README.md [hunk 2]
```diff
@@ -112,6 +127,30 @@ return createRouter(
...
+
+The `entityValidator` is a function you can pass into the processor which determines which entities should have language tags fetched from linguist and added to the entity. By default, this will only run on entities of `kind: Component`, however this function let's you fully customize which entities should be processed, for example by checking existence of certain annotations.
+
+### tagTransformer
+
+The `tagTransformer` option allows you to supply a custom function that can transform a linguist supplied language tag before it is appended to the entity tags. You can also use it to ignore certain language tags entirely by transforming the tag to `undefined`.
+
+### cacheTTL
+
+The `cacheTTL` option allows you to determine for how long this processor will cache languages for an entityRef before refreshing from the linguist backend. As this processor will run continuously, this cache is supplied to limit the load done on the linguist DB and API.
```
Comment (awanlin) [inline plugins/linguist-backend/README.md:144]: Adding backticks (`) around entityRef will make Vale happy

File: plugins/linguist-backend/README.md [hunk 3]
```diff
@@ -112,6 +127,30 @@ return createRouter(
...
 
+## Tags Processor Options
+
+The `LinguistTagsProcessor` is also configurable with the following configuration options
+
+### entityValidator
+
+The `entityValidator` is a function you can pass into the processor which determines which entities should have language tags fetched from linguist and added to the entity. By default, this will only run on entities of `kind: Component`, however this function let's you fully customize which entities should be processed, for example by checking existence of certain annotations.
+
+### tagTransformer
```
Comment (awanlin) [inline plugins/linguist-backend/README.md:138]: Adding backticks (`) around tagTransformer will make Vale happy

File: plugins/linguist-backend/README.md [hunk 4]
```diff
@@ -112,6 +127,30 @@ return createRouter(
 
 **Note:** This has the potential to cause a lot of processing, be very thoughtful about this before hand
 
+## Tags Processor Options
+
+The `LinguistTagsProcessor` is also configurable with the following configuration options
+
+### entityValidator
+
+The `entityValidator` is a function you can pass into the processor which determines which entities should have language tags fetched from linguist and added to the entity. By default, this will only run on entities of `kind: Component`, however this function let's you fully customize which entities should be processed, for example by checking existence of certain annotations.
```
Comment (awanlin) [inline plugins/linguist-backend/README.md:136]: Can we get an example?

File: plugins/linguist-backend/README.md [hunk 5]
```diff
@@ -112,6 +127,30 @@ return createRouter(
...
+
+### tagTransformer
+
+The `tagTransformer` option allows you to supply a custom function that can transform a linguist supplied language tag before it is appended to the entity tags. You can also use it to ignore certain language tags entirely by transforming the tag to `undefined`.
```
Comment (awanlin) [inline plugins/linguist-backend/README.md:140]: Same here, an example would be great!

File: plugins/linguist-backend/README.md [hunk 6]
```diff
@@ -112,6 +127,30 @@ return createRouter(
...
+
+### cacheTTL
+
+The `cacheTTL` option allows you to determine for how long this processor will cache languages for an entityRef before refreshing from the linguist backend. As this processor will run continuously, this cache is supplied to limit the load done on the linguist DB and API.
+
+By default, this processor will cache languages for 30 minutes before refreshing from the linguist database.
+
+You can optionally disable the cache entirely by passing in a cacheTTL duration of 0 minutes.
```
Comment (awanlin) [inline plugins/linguist-backend/README.md:148]: An example for this too, please

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+      ...this.loggerMeta,
+      entityRef,
+    });
+
+    const languages =
+      this.cacheTTLMilliseconds > 0
+        ? await this.getCachedLanguages(cache, entityRef)
+        : await this.getLanguagesFromLinguistAPI(entityRef);
+
+    const tags = (entity.metadata.tags ||= []);
```
Comment (awanlin) [inline plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:215]: More a general question, I haven't seen this before but is this the same as `const tags = (entity.metadata.tags || []);`? That `=` after the `||` is throwing me off

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    const originalTagCount = tags.length;
+
+    languages.forEach(lang => {
+      if (!tags.includes(lang)) {
+        const cleanedUpLangTag = this.tagTransformer(lang);
```
Comment (awanlin) [inline plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:220]: If I'm reading this correctly if they don't pass a `tagTranformer` function then the languages as used as is from Linguist? Won't that be problematic as the languages can have spaces in their names and I know the tags in Backstage won't allow that by default. Maybe we should have a default for this?

Here's the list of languages Linguist uses: https://github.com/github-linguist/linguist/blob/master/lib/linguist/languages.yml

**Review by awanlin — CHANGES_REQUESTED** [review]
This is really awesome work @bforbis, left some comments as there are a few things we should consider. I'm pretty picky about the `README` as it will save us a lot of questions as people use this.

Also, it would be great to have some tests around this as well. I've tried to get most of the core features tested so would good so see that done for this as well

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 7 above)
Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:215]: Yeah, this line is using some trickery.

The `||=` will just set the `tags` property on `entity.metadata` if it doesn't exist already. And then I'm assigning to a local variable so that it's less typing to push data into the tags.

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 8 above)
Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:220]: good idea, i haven't tested all of these.

Comment (awanlin) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:220]: I mean, don't feel like you need to test all of them but if we sample from that list and pull out the odd ones to have in automated tests that would be good

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 7 above)
Comment (awanlin) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:215]: Cool, learning things here, thanks! Is there a name for this so I can read up on it a bit more?

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 8 above)
Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:220]: Just tested a few with spaces and special characters, I didn't see any problems. So not sure if there is anything to do here.

image

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 7 above)
Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:215]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR_assignment

**Comment by bforbis** [comment]
I've re-worked the README per the code comments. I'll take a pass later to add tests if everything else looks good.

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+   * attached to an entity. If the tag transformer emits an "undefined" type, then it
+   * will not be added as a tag to the entity.
+   */
+  tagTransformer?: TagTransformer;
+  /**
+   * A function which determines which entities should be processed by the LinguistTagProcessor.
+   *
+   * The default is to process all entities of kind=Component
+   */
+  entityValidator?: EntityValidator;
```
Comment (freben) [inline plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:65]: Wouldn't call it a validator. That's what we call those that throw if a thing is malformed. `shouldProcessEntity` perhaps? Explains what it does

**Review by freben — COMMENTED** [review]
Just some random comments

**Comment by freben** [comment]
I see that vale is spell checking `lanugages` for you.

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 8 above)
Comment (awanlin) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:213]: @freben, I'm happy that the above works but why does it work? The docs say: "Each tag must be sequences of [a-z0-9] separated by -, at most 63 characters in total." (https://backstage.io/docs/features/software-catalog/descriptor-format#tags-optional) Are the docs wrong and need to be updated?

**Review by awanlin — APPROVED** [review]

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 8 above)
Comment (pjungermann) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:213]: from experience, the mentioned rules get applied for tags -- at least, when provided in catalog files. I wonder why it worked here 🤔

Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:213]: Could this be an "order of processors" problem? Maybe the tags don't get validated since this gets added to the catalog builder at a later stage?

Comment (pjungermann) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:213]: @awanlin the docs seem not the be up-to-date according to some code I found

https://github.com/backstage/backstage/blob/master/packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.ts#L82
```ts
          case 'isValidTag':
            expectation =
              'a string that is sequences of [a-z0-9+#] separated by [-], at most 63 characters in total';
            break;
```

https://github.com/backstage/backstage/blob/master/packages/catalog-model/src/entity/policies/FieldFormatEntityPolicy.ts#L148
```ts
    const tags = entity.metadata.tags ?? [];

    for (let i = 0; i < tags.length; ++i) {
      require(`tags.${i}`, tags[i], this.validators.isValidTag);
    }
```

https://github.com/backstage/backstage/blob/master/packages/catalog-model/src/validation/makeValidator.ts#L29
```ts
  isValidTag: (value: unknown): boolean => {
    // NOTE(freben): This one is a bit of an oddball and doesn't fit well anywhere to delegate to, so it's just inlined for now.
    return (
      typeof value === 'string' &&
      value.length >= 1 &&
      value.length <= 63 &&
      /^[a-z0-9:+#]+(\-[a-z0-9:+#]+)*$/.test(value)
    );
  },
```

Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:213]: aha! If I switch this from post-process to pre-process, I will see validation errors.

image

Comment (freben) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:213]: Yeah entity mutations should go in pre. It's a bit unfortunate that it's technically possible to break contract in post...

Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:213]: I realize this puts this PR into a bit of a bind, but perhaps this is a use case for [Changing the validation rules for the entity envelope](https://backstage.io/docs/features/software-catalog/extending-the-model#changing-the-validation-rules-for-the-entity-envelope-or-metadata-fields)?

This is a plugin after all, so integrators are opting into this functionality. The setup steps could involve also relaxing the tag restriction rules, or we could just keep this as a post-process phase to ignore the restrictions.

Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:213]: Ok, looking into this some more perhaps it's not as bad as I thought.

I tested the entire linguist data set using the following script

```js
const yaml = require('js-yaml');
const fs = require('fs')

async function main() {
  const langs = yaml.load(fs.readFileSync('./config.yaml'));

  const valid = [];
  const invalid = []
  const validCleaned = [];
  const invalidCleaned = [];

  Object.keys(langs).forEach(lang => {
    if ( langs[lang].type === 'programming' ) {
      if (isValidTag(lang)) {
        valid.push(lang);
      }
      else {
        invalid.push(lang);
      }
      const cleaned = cleanTag(lang);
      if (isValidTag(cleaned)) {
        validCleaned.push(cleaned);
      }
      else {
        invalidCleaned.push(cleaned);
      }
    }
  });

  const results = {
    valid: valid.length,
    invalid: invalid.length,
    validCleaned: validCleaned.length,
    invalidCleaned: invalidCleaned.length
  }

  console.log(JSON.stringify(results, null, 2));
  console.log(invalidCleaned);
}

function isValidTag (value) {
  // NOTE(freben): This one is a bit of an oddball and doesn't fit well anywhere to delegate to, so it's just inlined for now.
  return (
    typeof value === 'string' &&
    value.length >= 1 &&
    value.length <= 63 &&
    /^[a-z0-9:+#]+(\-[a-z0-9:+#]+)*$/.test(value)
  );
}

function cleanTag (value) {
  return value
    .replace(/\s+/g, '-')
    .toLowerCase();
}

main();
```

**Results**
```
{
  "valid": 9,
  "invalid": 457,
  "validCleaned": 459,
  "invalidCleaned": 7
}
[
  'asp.net',
  "cap'n-proto",
  'f*',
  'peg.js',
  "ren'py",
  'visual-basic-.net',
  'visual-basic-6.0'
]
```

Most of the issues are the fact that capital letters and spaces are not allowed in tags. I think asp.net is the only language in the short list of "still invalid" language tags that may just need a custom mapping. So that could be hard-mapped to "asp-dot-net" and the rest I think can be ignored.

Thoughts?

Comment (pjungermann) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:213]: you can use the invert to catch them all. And be aware that the first char is not allowed to be `-`.

```ts
value.replace(/^[^a-z0-9:+#]/, '').replace(/[^a-z0-9:+#-]+/g, '-')
```

Comment (pjungermann) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:213]: re your question:
In my opinion, it should be a preprocessing step that follows the restrictions for tags. Esp. if it is part of the main repo and not a separate plugin. Requiring to change the rules for the tags field will reduce the usefulness/ability to adopt this processor.

Comment (pjungermann) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:213]: As the docs don't match the actually allowed chars, I just created PR #18364 to fix this.

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+    linguistApi.searchParams.append('entityRef', entityRef);
+    const linguistData = await fetch(linguistApi).then(
+      res => res.json() as Promise<Languages>,
+    );
+    if (!linguistData || !linguistData.processedDate) {
+      return [];
+    }
+
+    return linguistData.breakdown
+      .filter(b => b.type === 'programming' && b.bytes > this.bytesThreshold)
```
[the remark this answers is not in the document]
Comment (pjungermann) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:164]: Currently, there are 
```'programming', 'data', 'markup', 'prose'```

`prose` (17)
```ts
['AsciiDoc', 'Creole', 'Gemini', 'Gettext Catalog', 'Markdown', 'Muse', 'Org', 'Pod', 'Pod 6', 'RDoc', 'RMarkdown', 'Texinfo', 'Text', 'Textile', 'Vim Help File', 'Wikitext', 'reStructuredText']
```

`markup` (60):
```ts
['API Blueprint', 'Antlers', 'Astro', 'BibTeX', 'Bikeshed', 'Blade', 'CSS', 'Closure Templates', 'D2', 'EJS', 'Ecmarkup', 'Genero Forms', 'HTML', 'HTML+ECR', 'HTML+EEX', 'HTML+ERB', 'HTML+PHP', 'HTML+Razor', 'Haml', 'Handlebars', 'Jinja', 'Jupyter Notebook', 'Kit', 'Latte', 'Less', 'Liquid', 'MDX', 'MTML', 'Marko', 'Mask', 'Mermaid', 'Mustache', 'Nunjucks', 'Pic', 'PostCSS', 'PostScript', 'Pug', 'RAML', 'RUNOFF', 'Rich Text Format', 'Riot', 'Roff', 'Roff Manpage', 'SCSS', 'SRecode Template', 'Sass', 'Scaml', 'Slim', 'StringTemplate', 'Stylus', 'SugarSS', 'Svelte', 'TeX', 'Tea', 'Twig', 'Velocity Template Language', 'Vim Snippet', 'Vue', 'YASnippet', 'kvlang']
```

`data` (158): includes many formats including Gradle, JSON, some entries related to OpenAPI Specification (OAS), PlaneUML, etc. (too many to list them all)

File: plugins/linguist-backend/README.md [hunk 11]
```diff
@@ -112,6 +113,101 @@ return createRouter(
...
+#### `bytesThreshold`
+
+The `bytesThreshold` option allows you to control a number of bytes threshold which must be surpassed before a language tag will be emitted by this processor. As an example, some repositories may have short build scripts written in Bash, but you may only want the main language of the project emitted (an alternate way to control this is to use the `tagTransformer` to reject any "bash" languages).
+
+```ts
+new LinguistLanguageTagsProcessor({
+  logger: env.logger,
+  discovery: env.discovery,
+  // Ignore languages with less than 5000 bytes in a repo.
+  bytesThreshold: 5000,
```
Comment (pjungermann) [inline plugins/linguist-backend/README.md:207]: It would be great to support options like that as configuration in `app-config.yaml` (e.g., also for `cacheTTL`)

**Review by pjungermann — COMMENTED** [review]
Looking forward to using this. I left a few suggestions and questions.

File: plugins/linguist-backend/README.md [hunk 12]
```diff
@@ -112,6 +113,101 @@ return createRouter(
...
+As an example, you may choose to extend this to support both `Component` and `Resource` kinds along with allowing an opt-out annotation on the entity which entity authors can use.
+
+```ts
+new LinguistLanguageTagsProcessor({
+  logger: env.logger,
+  discovery: env.discovery,
+  shouldProcessEntity: (entity: Entity) => {
+    if (
+      ['Component', 'Resource'].includes(entity.kind) &&
+      !entity.metadata.annotations?.['backstage.io/ignore-linguist-tags']
```
[the remark this answers is not in the document]
Comment (bforbis) [reply plugins/linguist-backend/README.md:152]: I can switch it to be a local annotation value that is obviously an example and change it to opt-in

File: plugins/linguist-backend/README.md [hunk 13]
```diff
@@ -112,6 +113,101 @@ return createRouter(
...
 
+## Linguist Tags Processor
+
+The `LinguistTagsProcessor` can be added into your catalog builder as a way to incorporate the language breakdown from linguist as `metadata.tags` on your entities. Doing so enables the ability to easily filter for entities in your catalog index based on the language of the source repository.
+
+### Processor Setup
+
+Setup the linguist tag processor in `packages/backend/src/plugins/catalog.ts`.
+
+```ts
```
[the remark this answers is not in the document]
Comment (bforbis) [reply plugins/linguist-backend/README.md:124]: I've never seen this "title" attribute before. Does that end up rendering? This seems to not follow the convention of the rest of the document, so I'm not sure about adding this.

Can you point to an example repo that uses "title" so I can see what it looks like?

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+   * attached to an entity. If the tag transformer emits an "undefined" type, then it
+   * will not be added as a tag to the entity.
+   */
+  tagTransformer?: TagTransformer;
+  /**
+   * A function which determines which entities should be processed by the LinguistTagProcessor.
+   *
+   * The default is to process all entities of kind=Component
+   */
+  shouldProcessEntity?: ShouldProcessEntity;
```
[the remark this answers is not in the document]
Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:65]: I chose to not inline them because they are used in several places throughout the file, so a type alias made sense so that the signature can be updated in just one place if needed. Is there a particular reason you prefer in-lining?

File: plugins/linguist-backend/README.md (hunk 11 above)
Comment (bforbis) [reply plugins/linguist-backend/README.md:207]: Do you mean instead or as an additional separate way to configure? I feel like splitting the config options between processor constructor args and config is not a great experience (there are two places to look at), and we do require processor constructor args for passing functions like `tagTransformer` and `shouldProcessEntity`.

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+   * language for the repo.
+   */
+  bytesThreshold?: number;
+}
+
+/**
+ * These are customized transformers for certain languages where
+ * regex is not good enough to clean them up.
+ */
+const CUSTOM_LANG_MAP: Record<string, string> = {
```
[the remark this answers is not in the document]
Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:85]: This feels like it should remain private to me. An implementers choice to build their own `tagTransformer` means they are taking complete control. We do expose this functionality itself already through the `defaultTagTransformer`

File: plugins/linguist-backend/README.md [hunk 16]
```diff
@@ -112,6 +113,101 @@ return createRouter(
...
+    return ['python', 'java', 'typescript'].includes(cleaned)
+      ? cleaned
+      : undefined;
+  },
+});
+```
+
+#### `cacheTTL`
+
+The `cacheTTL` option allows you to determine for how long this processor will cache languages for an `entityRef` before refreshing from the linguist backend. As this processor will run continuously, this cache is supplied to limit the load done on the linguist DB and API.
```
[the remark this answers is not in the document]
Comment (bforbis) [reply plugins/linguist-backend/README.md:184]: I think some more research is needed before determining whether the cache is needed / not needed. I have not run this processor at large scale, and will admit that I created this cache to be a bit defensive in the situation of large catalogs that are constantly churning their entities. Each run of the processor loop would be fetching individual language breakdowns for every component entity, which is a lot of extra db reads for unchanging data.

In a 30 min default, I was trying to balance eventual consistency for relieving reads on DBs, but obviously that assumes that the reads are expensive enough in the first place to warrant that caching complexity and spend of extra process memory footprint (I assume, fairly small here as each entity probably only has a handful of languages associated, we may be talking MBs here).

>Values that are set during a processing run will only be visible in the directly
following run. The cache will be overwritten every run unless no new cache items
are written, in which case the existing values remain in the cache.

I'd like to know more about the above. What is the intention behind "values will only be visible in the directly following run"? Is this to make sure that caches are guaranteed to be bust to prevent growing memory? If I made sure to explicitly call `cache.set()` on each run with the previous cached values, would i be able to perpetuate the cache across several runs?

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts]: added support for this

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+ * Converts language tags from linguist to something acceptable by
+ * the tag requirements for backstage
+ *
+ * @param tag - A language tag from linguist
+ * @returns Cleaned up language tag
+ * @public
+ */
+export const defaultTagTransformer: TagTransformer = (tag: string) => {
+  const cleaned = tag.replace(/\s+/g, '-').toLowerCase();
+  return CUSTOM_LANG_MAP[cleaned] || cleaned;
```
[the remark this answers is not in the document]
Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:103]: I intentionally chose to use `||` operator here. I don't see a point in building a custom map to any falsey value, that would be an error in my mind.

As for checking the value in the map before doing the transforms, I don't have a strong opinion one way or another. As it is now, the keys in the `CUSTOM_LANG_MAP` are already transformed, which is why the mapping is `asp.net => asp-dot-net`. So in this case, it is correct to pass in `CUSTOM_LANG_MAP[cleaned]`.

I do wonder if there should just be a global replace of `.` to `-dot-`. In your example of replacers above, they just get stripped out entirely. Is that better?

**Comment by awanlin** [comment]
Hi @bforbis, just wanted to say thanks for putting in all the effort into this. I think in the end it's going to be a pretty awesome feature! 

@freben and @pjungermann thanks for the reviews as well!

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (pjungermann) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:65]: I thought about the API surface of the plugin / the API report. If they are "public", it is harder to modify or remove them later without a _potential_ breaking change.

File: plugins/linguist-backend/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (pjungermann) [reply plugins/linguist-backend/README.md:124]: sure, see e.g., https://backstage.io/docs/integrations/github/discovery

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (pjungermann) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:103]: sure, makes sense to keep `||` then. Just wanted to confirm.

Re `CUSTOM_LANG_MAP[cleaned]`, you are right, your code worked. I just misread/thought you would replace more already.

However, I would still recommend sticking to the original value instead of the already mutated one.
You can use the actual language names from linguist as keys then. To me that seems far more easier to maintain and understand.

In this case, `ASP.NET` instead of `asp.net`.

Also, the static mapping makes the standard transformation/cleaning unnecessary. And we can even skip the execution of the regex replace operation(s).

As there is just one exception at the moment, the difference will be subtle, though.

[the remark this answers is not in the document]
Comment (pjungermann) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:103]: > In your example of replacers above, they just get stripped out entirely. Is that better?

As long as there is no other mapping, yes, they would get removed. All invalid chars will be removed/replaced. They are not allowed characters. If special handling for certain chars is desired, it could be added on top.


> I do wonder if there should just be a global replace of . to -dot-. 

So far, there are the following ones:
```ts
['ASN.1', 'ASP.NET', 'Gemfile.lock', 'PEG.js', 'Visual Basic .NET', 'Visual Basic 6.0', 'robots.txt']
```

if we apply this as general rule, we end up with
```ts
['asn-dot-1', 'asp-dot-net', 'gemfile-dot-lock', 'peg-dot-js', 'visual-basic-dot-net', 'visual-Basic-6-dot-0', 'robots-dot-txt']
```
vs
```ts
['asn-1', 'asp-net', 'gemfile-lock', 'peg-js', 'visual-basic-net', 'visual-basic-6-0', 'robots-txt']
```

Personally, I think it only makes sense for `ASP.NET` and probably also `Visual Basic .NET`, at which you speak `.` as `dot` as well.

Basically, `.NET` as `-dot-net` could make sense.

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (pjungermann) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:89]: I might have been not very clear. Just to be sure we are on the same page:
I meant to allow users to provide their custom mapping. Maybe as an extension to the default one. So, not the field itself but more the feature.

File: plugins/linguist-backend/README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (pjungermann) [reply plugins/linguist-backend/README.md:184]: re whether caching is needed or not. I mainly wanted to raise the question of whether we really need it. @awanlin maybe you can give a better opinion as the author of the linguist plugins.

File: plugins/linguist-backend/README.md (hunk 11 above)
Comment (pjungermann) [reply plugins/linguist-backend/README.md:207]: in general, I prefer a config-driven vs code-driven approach. You could still allow both if you like. I would also differentiate between the functions you can pass and the other configuration options. Those are rather extension points you provide.

In general, Backstage moves in the direction of declarative integration.

I guess, it would be good if you can add support for the new `@backstage/backend-plugin-api`. You can find examples if you search for `createBackendModule` (which would fit to your processor) and `createBackendPlugin`.

Not sure if this was done for the linguist-backend already or not.

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts]: I took your suggestions and am only replacing a custom rule for `.net` specifically. I think that makes sense.

I also took a slightly different approach with regex which I think is a bit cleaner / easier to read / and handles cases of double slashes.

Also added tests across the full linguist data-set, including a jest snapshot which can now be easily reviewed when making transformation changes / linguist adding new supported languages.

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts:65]: Hmm, wouldn't changing the in-line signature or fieldName of the option be just as breaking as changing anything on a type alias? Not sure I follow what the difference is. I guess the only difference is that you may change the type name and not the signature?

File: plugins/linguist-backend/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (bforbis) [reply plugins/linguist-backend/README.md]: @pjungermann I just tried pushing this change and it does not render in github. The site you ended up linking is the main backstage documentation site which uses a micro-site pre-processor. As this is standard github Markdown, I don't think we can use this feature

File: plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (bforbis) [reply plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts]: This is now exposed

File: plugins/linguist-backend/README.md (hunk 11 above)
Comment (bforbis) [reply plugins/linguist-backend/README.md]: I've added support for `app-config.yaml`.

**Comment by pjungermann** [comment]
@bforbis Should we a re-review? (you can request that at the top) Or are you planning more changes?

**Comment by awanlin** [comment]
@bforbis what's left on this? @freben any chance you can take another look please?

**Comment by bforbis** [comment]
Hi @awanlin, sorry lots of vacation. I don't plan to make more changes unless there are stronger opinions on the caching.

File: plugins/linguist-backend/src/index.ts [hunk 18]
```diff
@@ -20,6 +20,7 @@
  * @packageDocumentation
  */
 
+export * from './processor';
```
Comment (freben) [inline plugins/linguist-backend/src/index.ts:23]: technically this could have gone in a new `plugins/catalog-backend-module-linguist`, but that can come in a future update down the line :)

**Review by freben — APPROVED** [review]
I say let's ship it and iterate!

**Comment by awanlin** [comment]
Thanks @freben, just what I was thinking, merge and build on it as needed! 🚀 

Many thanks for this contribution @bforbis, this is awesome work and feel like many in the Community will be using this 🎉  

Thanks as well to @pjungermann for the reviews 👍

**Comment by freben** [comment]
@awanlin <3 follow-up though! https://github.com/backstage/backstage/pull/18938
